superset-frontend/src/components/Pagination/index.ts

```diff
@@ -0,0 +1,26 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { Next } from './Next';
+import { Prev } from './Prev';
+import { Item } from './Item';
+import { Ellipsis } from './Ellipsis';
+import { PaginationWrapper } from './PaginationWrapper';
```

Comment:
I think you can just `export default Pagination` here and do assign `Next, Prev, Item, Ellipsis` to the functional component, as the original component name `Pagination` is more natural than `PaginationWrapper`.

```ts
type PaginationChild = Next | Prev | Item | Ellipsis;

function Pagination({
  children
} : {
  children: PaginationChild | PaginationChild[]
}) {
  return <PaginationList role="navigation">{children}</PaginationList>;
}

Pagination.Next = Next;
Pagination.Prev = Prev;
Pagination.Item = Item;
Pagination.Ellipsis = Ellipsis;

export default Pagination;
```

Reply:
I chose this way because it is easier to understand the code. Many develops have difficulty understanding different types for `children`.
And doing assignments like that is not the kind of thing that someone catches his eye and understands. (And it is more difficult to get to the "original component").
To tell you the truth, I only see disadvantages in using the way you proposed...

Reply:
@ktmud Do you agree or still want me to make the changes?

Reply:
Adding typing to `children` was just another suggestion unrelated to code organization. Having custom types for `children` enforces the usage of `<Pagination>...</Pagination>` so that TypeScript can warn you when other children types are used. E.g.  this code

```tsx
<Pagination>
  <li className="pagi">Select control</li>
</Pagination>
```

should show an error because an arbitrary `childNode` will either render incorrectly or create customization outside of the component itself, defeating the purpose of component abstraction.

Re: `PaginationWrapper` vs `Pagination`, it really doesn't matter that much. But when the component is used outside of `createUltimatePagination`, do you really think

```ts
import Pagination from 'src/components/pagination';

<Pagination.PaginationWrapper>
  <Pagination.Prev />
  <Pagination.Prev />
</Pagination.PaginationWrapper>
```

reads better than

```ts
import Pagination from 'src/components/pagination';

<Pagination>
  <Pagination.Prev />
  <Pagination.Prev />
</Pagination>
```

?

I don't get your point about assignments.

I'll defer to @nytai on which style to use since he's the original author of this component. TBH the original class component style also works perfectly fine for me.

As a general note, if all we wanted was adding tests to make sure components work correctly, I'd refrain from doing pure code style refactoring unless there are clear benefits to do so.

Reply:
> should show an error because an arbitrary `childNode` will either render incorrectly or create customization outside of the component itself, defeating the purpose of component abstraction.

I agree, that really makes sense.

Reply:
> Re: `PaginationWrapper` vs `Pagination`, it really doesn't matter that much. But when the component is used outside of `createUltimatePagination`, do you really think

The second example is really better but at the same time I am bothered to put something like "Pagination.Next = Next" because in my opinion this would need to be tested as well (which is an option).
This code also doesn't seem like something that a person with little experience will understand, and it is also more "difficult" to find the original component. I understand that there are benefits but the bad part seems to me to be greater than the benefit.

Perhaps changing `PaginationWrapper` to `Container` will improve that part, like this:
```ts
import Pagination from 'src/components/pagination';

<Pagination.Container>
  <Pagination.Prev />
  <Pagination.Prev />
</Pagination.Container>
```
It's not that good but it's close...

Reply:
> I'll defer to @nytai on which style to use since he's the original author of this component. TBH the original class component style also works perfectly fine to me.

I agree, it makes sense to me.

Reply:
> As a general note, if all we wanted was adding tests to make sure components work correctly, I'd refrain from doing pure code style refactoring unless there are clear benefits to do so.

I agree that it works. However, changing classes to functions improves reading, facilitates maintenance, facilitates testing, code minification (and in some cases performance too). I personally avoid class components as much as possible so I think that changing from class to function is always an improvement.
But I understand your point

Reply:
> Perhaps changing PaginationWrapper to Container will improve that part, like this:

We see fairly often the `<Component><Component.Child /></Component>` pattern in Superset (probably a byproduct of AntD), e.g. `Menu` and `Menu.Item`, `Panel` and `Panel.Heading`, `Panel.Body`, so I really don't see the need of breaking the convention here. I'm not sure why it is difficult to understand what the assignment statement does. Extra test is not needed if you just use `Pagination.Next` etc in the tests as well.

If you think the code is hard to understand to new comers, maybe add a comment like this:

```ts
// Expose child components as component attribute to enforce namespacing
// and avoid the need to import them separately.
Pagination.Next  = Next;
Pagination.Prev  = Prev;
```

Reply:
I have no strong opinions here. Some history: this was implemented as a quickfix once the previous component broke after we upgraded react-bootstrap. The api was designed to mirror the newer react bootstrap pagination api https://react-bootstrap.github.io/components/pagination/ and it's also in line with the compound component pattern heavily used by AntD.

Reply:
@ktmud 
> ```ts
> type PaginationChild = Next | Prev | Item | Ellipsis;
> ```

I tried to do this but it gives this error.
<img width="939" alt="Captura de Tela 2021-02-24 às 9 12 36 AM" src="https://user-images.githubusercontent.com/1014611/108999142-a7521d00-7680-11eb-9a92-96b7561541fb.png">

Putting the typeof looks like this:
<img width="726" alt="Captura de Tela 2021-02-24 às 9 15 08 AM" src="https://user-images.githubusercontent.com/1014611/108999305-de283300-7680-11eb-90a8-593d1900c0a1.png">

I believe it is because we are working with functions here...For things like these, I think that typing in children usually gets in the way of helping. Do you think it would be interesting to keep the type?

Reply:
> ```ts
> // Expose child components as component attribute to enforce namespacing
> // and avoid the need to import them separately.
> Pagination.Next  = Next;
> Pagination.Prev  = Prev;
> ```

@ktmud Still, done

Reply:
This should work:

```ts
import { ReactElement } from 'react';

type PaginationChild =
  | ReactElement<NextProps, typeof Next>
  | ReactElement<PrevProps, typeof Props>
```

But feel free to skip this change it it's too much work.

Reply:
> ```ts
> import { ReactElement } from 'react';
> 
> type PaginationChild =
>   | ReactElement<NextProps, typeof Next>
>   | ReactElement<PrevProps, typeof Props>
> ```

I tried this but it really didn't look very readable...
I ended up doing it like this:
```ts
interface PaginationProps {
  children: JSX.Element | JSX.Element[];
}
```
It is certainly not the best way possible, but it is easy to understand.

Reply:
Yeah, but the only thing this prevents is empty or text child nodes as it doesn't check props. Since this component normally isn't directly used, I don't think this matters that much so I'm OK with either way.

Just an idea for thought on how you can enforce children type in the future.